File: maze_game.js
// Maze Game Implementation in JavaScript

class MazeGame {
    // Action constants
    static UP = 0;
    static DOWN = 1;
    static LEFT = 2;
    static RIGHT = 3;
    static ACTIONS = [MazeGame.UP, MazeGame.DOWN, MazeGame.LEFT, MazeGame.RIGHT];

    // Cell types
    static EMPTY = 0;
    static WALL = 1;
    static COIN = 2;
    static EXIT = 3;

    constructor(mazeSize = {height: 10, width: 10}, numCoins = 5) {
        this.height = mazeSize.height;
        this.width = mazeSize.width;
        this.numCoins = numCoins;
        this.reset();
    }

    reset() {
        // Create maze
        this.maze = this._createMaze();

        // Place player at top-left area
        this.playerPos = {row: 1, col: 1};

        // Place exit at bottom-right area
        this.exitPos = {row: this.height - 2, col: this.width - 2};
        this.maze[this.exitPos.row][this.exitPos.col] = MazeGame.EXIT;

        // Place NPC far from player for easier gameplay
        this.npcPos = this._getRandomEmptyPos(8);

        // Place coins
        this.coins = [];
        for (let i = 0; i < this.numCoins; i++) {
            const coinPos = this._getRandomEmptyPos();
            this.coins.push(coinPos);
            this.maze[coinPos.row][coinPos.col] = MazeGame.COIN;
        }

        this.score = 0;
        this.steps = 0;
        this.maxSteps = 500;
        this.caught = false;
        this.won = false;

        // Track previous position to penalize staying in place
        this.prevPlayerPos = {...this.playerPos};

        // Track visit counts for each position to encourage exploration
        this.visitCounts = {};
        this._incrementVisitCount(this.playerPos);

        return this.getState();
    }

    _createMaze() {
        // Initialize empty maze
        const maze = Array(this.height).fill(0).map(() => Array(this.width).fill(MazeGame.EMPTY));

        // Border walls
        for (let i = 0; i < this.width; i++) {
            maze[0][i] = MazeGame.WALL;
            maze[this.height - 1][i] = MazeGame.WALL;
        }
        for (let i = 0; i < this.height; i++) {
            maze[i][0] = MazeGame.WALL;
            maze[i][this.width - 1] = MazeGame.WALL;
        }

        // Add internal walls (reduced density for more escape routes)
        if (this.width > 6) {
            for (let i = 2; i < this.height - 2; i += 3) {
                for (let j = 1; j < this.width / 2; j++) {
                    if (Math.random() < 0.25) {  // Reduced from 0.6 to 0.25 for more open space
                        maze[i][j * 2] = MazeGame.WALL;
                    }
                }
            }
        }

        if (this.height > 6) {
            for (let i = 1; i < this.height / 2; i++) {
                for (let j = 2; j < this.width - 2; j += 3) {
                    if (Math.random() < 0.25) {  // Reduced from 0.6 to 0.25 for more open space
                        maze[i * 2][j] = MazeGame.WALL;
                    }
                }
            }
        }

        return maze;
    }

    _getRandomEmptyPos(minDistanceFromPlayer = 0) {
        while (true) {
            const row = Math.floor(Math.random() * (this.height - 2)) + 1;
            const col = Math.floor(Math.random() * (this.width - 2)) + 1;

            if (this.maze[row][col] === MazeGame.EMPTY) {
                if (minDistanceFromPlayer > 0) {
                    const distance = Math.abs(row - this.playerPos.row) +
                                   Math.abs(col - this.playerPos.col);
                    if (distance >= minDistanceFromPlayer) {
                        return {row, col};
                    }
                } else {
                    return {row, col};
                }
            }
        }
    }

    getState() {
        return {
            playerPos: {...this.playerPos},
            npcPos: {...this.npcPos},
            coins: this.coins.map(c => ({...c})),
            score: this.score,
            steps: this.steps,
            caught: this.caught,
            won: this.won,
            maze: this.maze.map(row => [...row])
        };
    }

    _incrementVisitCount(pos) {
        const key = `${pos.row},${pos.col}`;
        if (!(key in this.visitCounts)) {
            this.visitCounts[key] = 0;
        }
        this.visitCounts[key]++;
        return this.visitCounts[key];
    }

    step(action) {
        if (this.caught || this.won) {
            return {state: this.getState(), reward: 0, done: true};
        }

        // Calculate distances before moving for progress rewards
        const prevCoinDist = this._getNearestCoinDistance(this.playerPos);
        const prevExitDist = this._getManhattanDistance(this.playerPos, this.exitPos);

        // Store previous position before moving
        this.prevPlayerPos = {...this.playerPos};

        // Move player
        const newPos = this._getNewPosition(this.playerPos, action);
        if (this._isValidMove(newPos)) {
            this.playerPos = newPos;
        }

        let reward = -0.05; // Reduced step penalty to encourage exploration

        // Penalize staying in the same position (didn't move)
        if (this.playerPos.row === this.prevPlayerPos.row &&
            this.playerPos.col === this.prevPlayerPos.col) {
            reward -= 1.0; // Strong penalty for hitting walls
        } else {
            // Track position visits with lighter penalty
            const visitCount = this._incrementVisitCount(this.playerPos);
            if (visitCount > 2) { // Only penalize after 2nd visit
                reward -= 0.1 * (visitCount - 2); // Lighter revisit penalty
            }
        }

        // Check coin collection
        let collectedCoin = false;
        for (let i = 0; i < this.coins.length; i++) {
            if (this.coins[i].row === this.playerPos.row &&
                this.coins[i].col === this.playerPos.col) {
                this.coins.splice(i, 1);
                this.maze[this.playerPos.row][this.playerPos.col] = MazeGame.EMPTY;
                this.score += 10;
                reward += 20; // Increased reward for collecting coin
                collectedCoin = true;
                break;
            }
        }

        // Check if reached exit
        if (this.playerPos.row === this.exitPos.row &&
            this.playerPos.col === this.exitPos.col) {
            this.won = true;
            // Bonus for winning with all coins collected
            const coinBonus = this.coins.length === 0 ? 50 : 0;
            // Bonus for winning quickly
            const efficiencyBonus = Math.max(0, (this.maxSteps - this.steps) * 0.1);
            reward += 150 + coinBonus + efficiencyBonus;
            return {state: this.getState(), reward: reward, done: true};
        }

        // Reward for making progress toward goals
        if (!collectedCoin && !(this.playerPos.row === this.prevPlayerPos.row &&
            this.playerPos.col === this.prevPlayerPos.col)) {
            // If there are coins left, reward getting closer to nearest coin
            if (this.coins.length > 0) {
                const newCoinDist = this._getNearestCoinDistance(this.playerPos);
                if (newCoinDist < prevCoinDist) {
                    reward += 0.5; // Reward for moving toward coin
                } else if (newCoinDist > prevCoinDist) {
                    reward -= 0.3; // Penalty for moving away from coin
                }
            }
            // If no coins left, reward getting closer to exit
            else {
                const newExitDist = this._getManhattanDistance(this.playerPos, this.exitPos);
                if (newExitDist < prevExitDist) {
                    reward += 1.0; // Strong reward for moving toward exit
                } else if (newExitDist > prevExitDist) {
                    reward -= 0.5; // Penalty for moving away from exit
                }
            }
        }

        // Move NPC
        this._moveNPC();

        // Check if caught
        if (this.playerPos.row === this.npcPos.row &&
            this.playerPos.col === this.npcPos.col) {
            this.caught = true;
            reward -= 50;
            return {state: this.getState(), reward: reward, done: true};
        }

        this.steps++;

        // Check max steps
        if (this.steps >= this.maxSteps) {
            reward -= 10;
            return {state: this.getState(), reward: reward, done: true};
        }

        return {state: this.getState(), reward: reward, done: false};
    }

    _getNewPosition(pos, action) {
        const newPos = {...pos};
        switch (action) {
            case MazeGame.UP:
                newPos.row--;
                break;
            case MazeGame.DOWN:
                newPos.row++;
                break;
            case MazeGame.LEFT:
                newPos.col--;
                break;
            case MazeGame.RIGHT:
                newPos.col++;
                break;
        }
        return newPos;
    }

    _isValidMove(pos) {
        if (pos.row < 0 || pos.row >= this.height ||
            pos.col < 0 || pos.col >= this.width) {
            return false;
        }
        return this.maze[pos.row][pos.col] !== MazeGame.WALL;
    }

    _getManhattanDistance(pos1, pos2) {
        return Math.abs(pos1.row - pos2.row) + Math.abs(pos1.col - pos2.col);
    }

    _getNearestCoinDistance(pos) {
        if (this.coins.length === 0) {
            return 999; // Large value if no coins
        }
        return Math.min(...this.coins.map(coin => this._getManhattanDistance(pos, coin)));
    }

    _moveNPC() {
        // NPC only moves 5% of the time to make it much easier to avoid and allow wins
        if (Math.random() > 0.05) {
            return;
        }

        // Random movement: NPC moves in random valid directions
        const validActions = [];

        for (const action of MazeGame.ACTIONS) {
            const newPos = this._getNewPosition(this.npcPos, action);
            if (this._isValidMove(newPos)) {
                validActions.push(action);
            }
        }

        // Move NPC to a random valid position
        if (validActions.length > 0) {
            const action = validActions[Math.floor(Math.random() * validActions.length)];
            this.npcPos = this._getNewPosition(this.npcPos, action);
        }
    }

    getStateRepresentation() {
        // Compact state representation for Q-learning
        const npcRelative = {
            row: this.npcPos.row - this.playerPos.row,
            col: this.npcPos.col - this.playerPos.col
        };

        const exitRelative = {
            row: this.exitPos.row - this.playerPos.row,
            col: this.exitPos.col - this.playerPos.col
        };

        let coinRelative = {row: 0, col: 0};
        if (this.coins.length > 0) {
            // Find nearest coin
            let minDist = Infinity;
            for (const coin of this.coins) {
                const dist = Math.abs(coin.row - this.playerPos.row) +
                           Math.abs(coin.col - this.playerPos.col);
                if (dist < minDist) {
                    minDist = dist;
                    coinRelative = {
                        row: coin.row - this.playerPos.row,
                        col: coin.col - this.playerPos.col
                    };
                }
            }
        }

        // Return string representation for dictionary key
        return JSON.stringify({
            player: this.playerPos,
            npc: npcRelative,
            exit: exitRelative,
            coin: coinRelative,
            coinsLeft: this.coins.length
        });
    }
}

File: maze_game.py
import numpy as np
import random
from typing import Tuple, List, Optional

class MazeGame:
    """
    Maze game environment with:
    - Player/AI agent that needs to reach the exit
    - NPC that chases the player
    - Coins to collect for score
    """

    # Action constants
    UP = 0
    DOWN = 1
    LEFT = 2
    RIGHT = 3
    ACTIONS = [UP, DOWN, LEFT, RIGHT]

    # Cell types
    EMPTY = 0
    WALL = 1
    COIN = 2
    EXIT = 3

    def __init__(self, maze_size: Tuple[int, int] = (10, 10), num_coins: int = 5):
        self.height, self.width = maze_size
        self.num_coins = num_coins
        self.reset()

    def reset(self) -> dict:
        """Reset the game to initial state"""
        # Create maze with walls
        self.maze = self._create_maze()

        # Place player at top-left area
        self.player_pos = [1, 1]

        # Place exit at bottom-right area
        self.exit_pos = [self.height - 2, self.width - 2]
        self.maze[self.exit_pos[0], self.exit_pos[1]] = self.EXIT

        # Place NPC at a random position (not too close to player)
        self.npc_pos = self._get_random_empty_pos(min_distance_from_player=4)

        # Place coins
        self.coins = []
        for _ in range(self.num_coins):
            coin_pos = self._get_random_empty_pos()
            self.coins.append(coin_pos)
            self.maze[coin_pos[0], coin_pos[1]] = self.COIN

        self.score = 0
        self.steps = 0
        self.max_steps = 500
        self.caught = False
        self.won = False

        return self._get_state()

    def _create_maze(self) -> np.ndarray:
        """Create a maze with walls around the border and some internal walls"""
        maze = np.zeros((self.height, self.width), dtype=int)

        # Border walls
        maze[0, :] = self.WALL
        maze[-1, :] = self.WALL
        maze[:, 0] = self.WALL
        maze[:, -1] = self.WALL

        # Add some internal walls to make it more interesting
        # Vertical walls
        if self.width > 6:
            for i in range(2, self.height - 2, 3):
                for j in range(1, self.width // 2):
                    if random.random() < 0.6:
                        maze[i, j * 2] = self.WALL

        # Horizontal walls
        if self.height > 6:
            for i in range(1, self.height // 2):
                for j in range(2, self.width - 2, 3):
                    if random.random() < 0.6:
                        maze[i * 2, j] = self.WALL

        return maze

    def _get_random_empty_pos(self, min_distance_from_player: int = 0) -> List[int]:
        """Get a random empty position in the maze"""
        while True:
            row = random.randint(1, self.height - 2)
            col = random.randint(1, self.width - 2)

            if self.maze[row, col] == self.EMPTY:
                if min_distance_from_player > 0:
                    distance = abs(row - self.player_pos[0]) + abs(col - self.player_pos[1])
                    if distance >= min_distance_from_player:
                        return [row, col]
                else:
                    return [row, col]

    def _get_state(self) -> dict:
        """Get current state representation"""
        return {
            'player_pos': tuple(self.player_pos),
            'npc_pos': tuple(self.npc_pos),
            'coins': [tuple(c) for c in self.coins],
            'score': self.score,
            'steps': self.steps,
            'caught': self.caught,
            'won': self.won,
            'maze': self.maze.copy()
        }

    def step(self, action: int) -> Tuple[dict, float, bool]:
        """
        Take a step in the environment
        Returns: (state, reward, done)
        """
        if self.caught or self.won:
            return self._get_state(), 0, True

        # Move player
        new_pos = self._get_new_position(self.player_pos, action)
        if self._is_valid_move(new_pos):
            self.player_pos = new_pos

        reward = -0.1  # Small negative reward for each step (encourages efficiency)

        # Check if player collected a coin
        if self.player_pos in self.coins:
            self.coins.remove(self.player_pos)
            self.maze[self.player_pos[0], self.player_pos[1]] = self.EMPTY
            self.score += 10
            reward += 10  # Reward for collecting coin

        # Check if player reached exit
        if self.player_pos == self.exit_pos:
            self.won = True
            reward += 100  # Big reward for winning
            return self._get_state(), reward, True

        # Move NPC (chases player)
        self._move_npc()

        # Check if NPC caught player
        if self.player_pos == self.npc_pos:
            self.caught = True
            reward -= 50  # Big penalty for getting caught
            return self._get_state(), reward, True

        self.steps += 1

        # Check if max steps reached
        if self.steps >= self.max_steps:
            reward -= 10  # Penalty for taking too long
            return self._get_state(), reward, True

        return self._get_state(), reward, False

    def _get_new_position(self, pos: List[int], action: int) -> List[int]:
        """Get new position based on action"""
        new_pos = pos.copy()
        if action == self.UP:
            new_pos[0] -= 1
        elif action == self.DOWN:
            new_pos[0] += 1
        elif action == self.LEFT:
            new_pos[1] -= 1
        elif action == self.RIGHT:
            new_pos[1] += 1
        return new_pos

    def _is_valid_move(self, pos: List[int]) -> bool:
        """Check if position is valid (not a wall and within bounds)"""
        row, col = pos
        if row < 0 or row >= self.height or col < 0 or col >= self.width:
            return False
        if self.maze[row, col] == self.WALL:
            return False
        return True

    def _move_npc(self):
        """Move NPC towards player using simple pathfinding"""
        # Simple chase AI: move towards player
        best_action = None
        best_distance = float('inf')

        for action in self.ACTIONS:
            new_pos = self._get_new_position(self.npc_pos, action)
            if self._is_valid_move(new_pos):
                # Calculate Manhattan distance to player
                distance = abs(new_pos[0] - self.player_pos[0]) + abs(new_pos[1] - self.player_pos[1])
                if distance < best_distance:
                    best_distance = distance
                    best_action = action

        # Move NPC with 80% probability (makes it slightly less perfect)
        if best_action is not None and random.random() < 0.8:
            self.npc_pos = self._get_new_position(self.npc_pos, best_action)

    def get_state_representation(self) -> np.ndarray:
        """
        Get a compact state representation for Q-learning
        Returns a tuple that can be used as a dictionary key
        """
        # Discretize relative positions
        player_pos = tuple(self.player_pos)
        npc_relative = (
            self.npc_pos[0] - self.player_pos[0],
            self.npc_pos[1] - self.player_pos[1]
        )
        exit_relative = (
            self.exit_pos[0] - self.player_pos[0],
            self.exit_pos[1] - self.player_pos[1]
        )

        # Get nearest coin relative position
        if self.coins:
            nearest_coin = min(self.coins,
                             key=lambda c: abs(c[0] - self.player_pos[0]) + abs(c[1] - self.player_pos[1]))
            coin_relative = (
                nearest_coin[0] - self.player_pos[0],
                nearest_coin[1] - self.player_pos[1]
            )
        else:
            coin_relative = (0, 0)

        return (player_pos, npc_relative, exit_relative, coin_relative, len(self.coins))

    def render_text(self):
        """Print text representation of the game"""
        print("\n" + "="*50)
        print(f"Score: {self.score} | Steps: {self.steps}")
        print("="*50)

        for i in range(self.height):
            for j in range(self.width):
                if [i, j] == self.player_pos:
                    print("P", end=" ")
                elif [i, j] == self.npc_pos:
                    print("N", end=" ")
                elif [i, j] == self.exit_pos:
                    print("E", end=" ")
                elif [i, j] in self.coins:
                    print("C", end=" ")
                elif self.maze[i, j] == self.WALL:
                    print("#", end=" ")
                else:
                    print(".", end=" ")
            print()

        if self.won:
            print("\n🎉 YOU WON! 🎉")
        elif self.caught:
            print("\n💀 CAUGHT BY NPC! 💀")
        print()
